src/common/mfu_daos.c

```diff
@@ -1313,3 +1330,2195 @@ int mfu_flist_copy_daos(
+                                        daos_handle_t cont,
+                                        uint64_t *total_size)
+{
+    int     rc = 0;
+    hid_t       status = 0;
+    char        *total_usr_attrs = "Total User Attributes";
+
+    /* record total number of user defined attributes, if it is 0,
+     * then we don't write any more attrs */
+    rc = daos_cont_list_attr(cont, NULL, total_size, NULL);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to retrieve user attributes on "
+                             "container");
+        goto out;
+    }
+
+    hdf5->attr_dims[0] = 1;
+    hdf5->attr_dtype = H5Tcopy(H5T_NATIVE_INT);
+    status = H5Tset_size(hdf5->attr_dtype, 8);
+    if (status < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create version attribute");
+        rc = 1;
+        goto out;
+    }
+    if (hdf5->attr_dtype < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create usr attr type");
+        rc = 1;
+        goto out;
+    }
+    hdf5->attr_dspace = H5Screate_simple(1, hdf5->attr_dims,
+                                         NULL);
+    if (hdf5->attr_dspace < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create version attribute dataspace");
+        rc = 1;
+        goto out;
+    }
+    hdf5->usr_attr_num = H5Acreate2(hdf5->file,
+                                    total_usr_attrs,
+                                    hdf5->attr_dtype,
+                                    hdf5->attr_dspace,
+                                    H5P_DEFAULT,
+                                    H5P_DEFAULT);
+    if (hdf5->usr_attr_num < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create version attribute");
+        rc = 1;
+        goto out;
+    }   
+    status = H5Awrite(hdf5->usr_attr_num, hdf5->attr_dtype,
+                      total_size);
+    if (status < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to write attribute");
+        rc = 1;
+        goto out;
+    }
+    status = H5Aclose(hdf5->usr_attr_num);
+    if (status < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to close attribute");
+        rc = 1;
+        goto out;
+    }
+out:
+    return rc;
+}
+
+static int cont_serialize_usr_attrs(struct hdf5_args *hdf5, daos_handle_t cont)
+{
+    int         rc = 0;
+    int         i = 0;
+    int         j = 0;
+    hid_t       status = 0;
+    uint64_t    total_size = 0;
+    uint64_t    size = 0;
+    uint64_t    cur = 0;
+    uint64_t    len = 0;
+    char        *buf = NULL;
+    char        **names = NULL;
+    int         num_attrs = 0;
+    void        **val_buf = NULL;
+    size_t      *sizes = NULL;
+
+    rc = cont_serialize_num_usr_attrs(hdf5, cont, &total_size);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to serialize number of user attrs: %d",
+                rc);
+        goto out;
+    }
+
+    /* skip serializing attrs if total_size is zero */
+    if (total_size == 0) {
+        goto out;
+    }
+
+    buf = malloc(total_size);
```

Comment:
This goes back to how much we want to make this code look like the rest of the mfu code, but we could switch to use the MFU_MALLOC wrappers like so:
```suggestion
    buf = MFU_MALLOC(total_size);
```
The MFU_MALLOC wrappers always return NULL when given a ``size==0``, so that we know by looking at the pointer value as to whether it is pointing to memory or not.

Reply:
I agree that we should use the MFU_MALLOC wrappers when possible, because they will log the file and line number for a failed malloc.